### PR TITLE
Pointer assignment semantic checks

### PR DESCRIPTION
**Pull request by tskeith**
See individual commits.

### CODE REVIEW

File: lib/evaluate/characteristics.cc [hunk 1]
```diff
@@ -636,7 +638,7 @@ std::optional<Procedure> Procedure::Characterize(
           [&](const semantics::ProcEntityDetails &proc)
               -> std::optional<Procedure> {
             if (symbol.attrs().test(semantics::Attr::INTRINSIC)) {
-              return intrinsics.IsUnrestrictedSpecificIntrinsicFunction(
+              return intrinsics.IsSpecificIntrinsicFunction(
```
Comment (klausler) [inline lib/evaluate/characteristics.cc]: Could you summarize for me what's changed here?  Do you permit procedure pointers to point to restricted specific intrinsic functions?

File: lib/semantics/check-declarations.cc [hunk 2]
```diff
@@ -659,6 +673,17 @@ bool CheckHelper::CheckDefinedAssignmentArg(
   return true;
 }
 
+// Report a conflicting attribute error if symbols has both of these attributes
```
Comment (klausler) [inline lib/semantics/check-declarations.cc]: s/has/have/

File: lib/evaluate/characteristics.cc (hunk 1 above)
Comment (tskeith) [reply lib/evaluate/characteristics.cc]: The change is to be able to detect specific intrinsic functions, both restricted and unrestricted. `IsSpecificIntrinsicFunction` returns a struct that has a flag to distinguish restricted and unrestricted. That flag is propagated to `SpecificIntrinsic` in `ProcedureDesignator` by `ExpressionAnalyzer::Designate`.

The new tests in `resolve46.f90` verify that procedure pointers can't point to restricted specific intrinsic functions.

File: lib/semantics/check-declarations.cc (hunk 2 above)
Comment (tskeith) [reply lib/semantics/check-declarations.cc]: Thanks
